File: kotlin/src/main/kotlin/ai/ensoul/sdk/resources/Personas.kt
package ai.ensoul.sdk.resources

import ai.ensoul.sdk.EnsoulHttpClient
import ai.ensoul.sdk.Page
import ai.ensoul.sdk.generated.PersonaBatchResponse
import ai.ensoul.sdk.generated.PersonaFiltersResponse
import ai.ensoul.sdk.generated.PersonaResponse
import ai.ensoul.sdk.generated.PersonalityVectorResponse
import io.ktor.client.statement.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonPrimitive

class Personas(private val client: EnsoulHttpClient) {

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun create(
        name: String,
        domain: String,
        personalityData: Map<String, Any?>? = null,
        extras: Map<String, Any?> = emptyMap(),
    ): PersonaResponse {
        val body = mutableMapOf<String, Any?>("name" to name, "domain" to domain)
        personalityData?.let { body["personality_data"] = it }
        body.putAll(extras.filterValues { it != null })
        val response = client.post("/v1/personas", json = body)
        return json.decodeFromString(PersonaResponse.serializer(), response.bodyAsText())
    }

    suspend fun get(personaId: String): PersonaResponse {
        val response = client.get("/v1/personas/$personaId")
        return json.decodeFromString(PersonaResponse.serializer(), response.bodyAsText())
    }

    suspend fun update(personaId: String, fields: Map<String, Any?> = emptyMap()): PersonaResponse {
        val body = fields.filterValues { it != null }
        val response = client.put("/v1/personas/$personaId", json = body)
        return json.decodeFromString(PersonaResponse.serializer(), response.bodyAsText())
    }

    suspend fun delete(personaId: String) {
        client.delete("/v1/personas/$personaId")
    }

    suspend fun list(
        page: Int = 1,
        perPage: Int = 20,
        region: String? = null,
        archetype: String? = null,
        country: String? = null,
        city: String? = null,
        extras: Map<String, Any?> = emptyMap(),
    ): Page<PersonaResponse> {
        val params = mutableMapOf<String, Any?>("page" to page, "per_page" to perPage)
        region?.let { params["region"] = it }
        archetype?.let { params["archetype"] = it }
        country?.let { params["country"] = it }
        city?.let { params["city"] = it }
        params.putAll(extras.filterValues { it != null })
        val response = client.get("/v1/personas", params = params)
        val text = response.bodyAsText()
        val data = json.parseToJsonElement(text).jsonObject
        val items = data["items"]!!.jsonArray.map {
            json.decodeFromJsonElement(PersonaResponse.serializer(), it)
        }
        return Page(
            items = items,
            total = data["total"]!!.jsonPrimitive.int,
            page = data["page"]!!.jsonPrimitive.int,
            perPage = data["per_page"]!!.jsonPrimitive.int,
            pages = data["pages"]!!.jsonPrimitive.int,
            client = client,
            method = "GET",
            path = "/v1/personas",
            params = params,
            deserializer = { json.decodeFromJsonElement(PersonaResponse.serializer(), it) },
        )
    }

    suspend fun batchCreate(
        personas: List<Map<String, Any?>>,
        batchId: String? = null,
        domain: String? = null,
    ): PersonaBatchResponse {
        val body = mutableMapOf<String, Any?>("personas" to personas)
        batchId?.let { body["batch_id"] = it }
        domain?.let { body["domain"] = it }
        val response = client.post("/v1/personas/batch", json = body)
        return json.decodeFromString(PersonaBatchResponse.serializer(), response.bodyAsText())
    }

    suspend fun getPersonality(personaId: String): PersonalityVectorResponse {
        val response = client.get("/v1/personas/$personaId/personality")
        return json.decodeFromString(PersonalityVectorResponse.serializer(), response.bodyAsText())
    }

    suspend fun getFilters(): PersonaFiltersResponse {
        val response = client.get("/v1/personas/filters")
        return json.decodeFromString(PersonaFiltersResponse.serializer(), response.bodyAsText())
    }

    suspend fun getConnections(personaId: String): List<JsonObject> {
        val response = client.get("/v1/personas/$personaId/connections")
        val text = response.bodyAsText()
        return json.parseToJsonElement(text).jsonArray.map { it.jsonObject }
    }
}

File: swift/Sources/Ensoul/Resources/Personas.swift
/// Personas resource for the Ensoul Swift SDK.
///
/// Wraps all `/v1/personas` endpoints. Every method is `async throws` since Swift
/// concurrency is async-only at the SDK boundary.
///
/// Example:
/// ```swift
/// let persona = try await client.personas.create(name: "Alice", domain: "acme")
/// let page    = try await client.personas.list(page: 1, perPage: 20)
/// for item in page.items { print(item.id) }
/// ```
import Foundation

// MARK: - Personas

@available(iOS 15.0, macOS 12.0, *)
public class Personas {
    private let client: HTTPClient

    public init(client: HTTPClient) {
        self.client = client
    }

    // MARK: - Create

    /// POST /v1/personas
    ///
    /// Creates a new persona in the given domain.
    public func create(
        name: String,
        domain: String,
        personalityData: [String: Any]? = nil,
        archetype: String? = nil,
        region: String? = nil,
        age: Int? = nil,
        country: String? = nil,
        city: String? = nil,
        backstory: String? = nil,
        coreValues: [String]? = nil,
        communicationStyle: [String: Any]? = nil
    ) async throws -> PersonaResponse {
        var body: [String: Any] = [
            "name": name,
            "domain": domain,
        ]
        if let personalityData { body["personality_data"] = personalityData }
        if let archetype { body["archetype"] = archetype }
        if let region { body["region"] = region }
        if let age { body["age"] = age }
        if let country { body["country"] = country }
        if let city { body["city"] = city }
        if let backstory { body["backstory"] = backstory }
        if let coreValues { body["core_values"] = coreValues }
        if let communicationStyle { body["communication_style"] = communicationStyle }

        let (data, _) = try await client.post("/v1/personas", body: body)
        let decoder = JSONDecoder()
        return try decoder.decode(PersonaResponse.self, from: data)
    }

    // MARK: - Get

    /// GET /v1/personas/{personaId}
    public func get(_ personaId: String) async throws -> PersonaResponse {
        let (data, _) = try await client.get("/v1/personas/\(personaId)")
        let decoder = JSONDecoder()
        return try decoder.decode(PersonaResponse.self, from: data)
    }

    // MARK: - Update

    /// PUT /v1/personas/{personaId}
    ///
    /// Updates an existing persona. Only provide the fields you want to change.
    public func update(
        _ personaId: String,
        name: String? = nil,
        personalityData: [String: Any]? = nil,
        archetype: String? = nil,
        region: String? = nil,
        age: Int? = nil,
        country: String? = nil,
        city: String? = nil,
        backstory: String? = nil,
        coreValues: [String]? = nil,
        communicationStyle: [String: Any]? = nil
    ) async throws -> PersonaResponse {
        var body: [String: Any] = [:]
        if let name { body["name"] = name }
        if let personalityData { body["personality_data"] = personalityData }
        if let archetype { body["archetype"] = archetype }
        if let region { body["region"] = region }
        if let age { body["age"] = age }
        if let country { body["country"] = country }
        if let city { body["city"] = city }
        if let backstory { body["backstory"] = backstory }
        if let coreValues { body["core_values"] = coreValues }
        if let communicationStyle { body["communication_style"] = communicationStyle }

        let (data, _) = try await client.put("/v1/personas/\(personaId)", body: body)
        let decoder = JSONDecoder()
        return try decoder.decode(PersonaResponse.self, from: data)
    }

    // MARK: - Delete

    /// DELETE /v1/personas/{personaId}
    public func delete(_ personaId: String) async throws {
        _ = try await client.delete("/v1/personas/\(personaId)")
    }

    // MARK: - List

    /// GET /v1/personas
    ///
    /// Returns a paginated page of personas. Use `page.nextPage()` to advance,
    /// or `for try await item in page.autoPagingSequence()` to iterate all items.
    public func list(
        page: Int = 1,
        perPage: Int = 20,
        domain: String? = nil,
        region: String? = nil,
        archetype: String? = nil,
        country: String? = nil,
        city: String? = nil
    ) async throws -> Page<PersonaResponse> {
        var params: [String: String] = [
            "page": String(page),
            "per_page": String(perPage),
        ]
        if let domain { params["domain"] = domain }
        if let region { params["region"] = region }
        if let archetype { params["archetype"] = archetype }
        if let country { params["country"] = country }
        if let city { params["city"] = city }

        let (data, _) = try await client.get("/v1/personas", params: params)
        return try Page.from(
            data: data,
            client: client,
            method: "GET",
            path: "/v1/personas",
            params: params
        )
    }

    // MARK: - Batch Create

    /// POST /v1/personas/batch
    ///
    /// Creates multiple personas in a single request.
    public func batchCreate(
        personas: [[String: Any]],
        batchId: String? = nil,
        domain: String? = nil
    ) async throws -> PersonaBatchResponse {
        var body: [String: Any] = ["personas": personas]
        if let batchId { body["batch_id"] = batchId }
        if let domain { body["domain"] = domain }

        let (data, _) = try await client.post("/v1/personas/batch", body: body)
        let decoder = JSONDecoder()
        return try decoder.decode(PersonaBatchResponse.self, from: data)
    }

    // MARK: - Get Personality

    /// GET /v1/personas/{personaId}/personality
    ///
    /// Returns the raw personality vector for a persona.
    public func getPersonality(_ personaId: String) async throws -> PersonalityVectorResponse {
        let (data, _) = try await client.get("/v1/personas/\(personaId)/personality")
        let decoder = JSONDecoder()
        return try decoder.decode(PersonalityVectorResponse.self, from: data)
    }

    // MARK: - Get Filters

    /// GET /v1/personas/filters
    ///
    /// Returns the set of valid filter values for listing personas.
    public func getFilters() async throws -> PersonaFiltersResponse {
        let (data, _) = try await client.get("/v1/personas/filters")
        let decoder = JSONDecoder()
        return try decoder.decode(PersonaFiltersResponse.self, from: data)
    }

    // MARK: - Get Connections

    /// GET /v1/personas/{personaId}/connections
    ///
    /// Returns the social graph connections for a persona as raw JSON dictionaries.
    /// The connection schema is domain-specific so raw `[String: Any]` is returned.
    public func getConnections(_ personaId: String) async throws -> [[String: Any]] {
        let (data, _) = try await client.get("/v1/personas/\(personaId)/connections")
        guard let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            throw EnsoulAPIError(
                statusCode: 200,
                error: "ParseError",
                message: "Expected JSON array for persona connections"
            )
        }
        return array
    }
}
